File: web/js/calorie-calculator.js
const CalorieCalculator = {
    // Mifflin-St Jeor BMR equation
    bmr(weightKg, heightCm, age, sex) {
        const base = (10 * weightKg) + (6.25 * heightCm) - (5 * age);
        return sex === 'male' ? base + 5 : base - 161;
    },

    // Total Daily Energy Expenditure
    tdee(bmr, activityMultiplier) {
        return bmr * activityMultiplier;
    },

    // Daily calorie target for weight loss
    dailyTarget(currentWeightLbs, goalWeightLbs, weeks, tdee) {
        const lbsToLose = currentWeightLbs - goalWeightLbs;
        if (lbsToLose <= 0 || weeks <= 0) return Math.round(tdee);

        const totalDeficit = lbsToLose * 3500; // 1 lb fat ≈ 3500 cal
        const dailyDeficit = totalDeficit / (weeks * 7);
        const target = tdee - dailyDeficit;
        return Math.max(1200, Math.round(target));
    },

    // Unit conversions
    lbsToKg(lbs) { return lbs * 0.453592; },
    inchesToCm(inches) { return inches * 2.54; },

    // Compute all values from a profile object
    computeAll(profile) {
        const weightKg = this.lbsToKg(profile.weightLbs);
        const heightCm = this.inchesToCm(profile.heightInches);
        const bmr = this.bmr(weightKg, heightCm, profile.age, profile.sex);
        const tdee = this.tdee(bmr, profile.activityMultiplier);
        const weeks = this.weeksUntil(new Date(profile.goalDate));
        const dailyTarget = this.dailyTarget(profile.weightLbs, profile.goalWeightLbs, weeks, tdee);

        return {
            bmr: Math.round(bmr),
            tdee: Math.round(tdee),
            dailyTarget,
            weeks,
            daysRemaining: this.daysUntil(new Date(profile.goalDate))
        };
    },

    weeksUntil(date) {
        const now = new Date();
        const diff = date.getTime() - now.getTime();
        return Math.max(1, Math.round(diff / (7 * 24 * 60 * 60 * 1000)));
    },

    daysUntil(date) {
        const now = new Date();
        now.setHours(0, 0, 0, 0);
        const target = new Date(date);
        target.setHours(0, 0, 0, 0);
        return Math.max(0, Math.round((target - now) / (24 * 60 * 60 * 1000)));
    }
};

File: CalorieTracker/Services/CalorieCalculator.swift
import Foundation

struct CalorieCalculator {
    /// Mifflin-St Jeor BMR equation
    static func bmr(weightKg: Double, heightCm: Double, age: Int, sex: Sex) -> Double {
        let base = (10 * weightKg) + (6.25 * heightCm) - (5 * Double(age))
        switch sex {
        case .male: return base + 5
        case .female: return base - 161
        }
    }

    /// Total Daily Energy Expenditure
    static func tdee(bmr: Double, activityLevel: ActivityLevel) -> Double {
        bmr * activityLevel.multiplier
    }

    /// Calculate daily calorie target to reach goal weight by goal date
    static func dailyTarget(
        currentWeightLbs: Double,
        goalWeightLbs: Double,
        weeks: Int,
        tdee: Double
    ) -> Int {
        let lbsToLose = currentWeightLbs - goalWeightLbs
        guard lbsToLose > 0, weeks > 0 else { return Int(tdee) }

        // 1 lb of fat ≈ 3500 calories
        let totalDeficit = lbsToLose * 3500
        let dailyDeficit = totalDeficit / Double(weeks * 7)

        // Cap deficit at reasonable level (no less than 1200 cal/day)
        let target = tdee - dailyDeficit
        return max(1200, Int(target))
    }

    /// Convenience: compute everything from a UserProfile
    static func dailyTargetForProfile(_ profile: UserProfile) -> Int {
        let bmrValue = bmr(
            weightKg: profile.weightKg,
            heightCm: profile.heightCm,
            age: profile.age,
            sex: profile.sex
        )
        let tdeeValue = tdee(bmr: bmrValue, activityLevel: profile.activityLevel)
        return dailyTarget(
            currentWeightLbs: profile.weightLbs,
            goalWeightLbs: profile.goalWeightLbs,
            weeks: profile.weeksToGoal,
            tdee: tdeeValue
        )
    }

    /// BMR for display
    static func bmrForProfile(_ profile: UserProfile) -> Int {
        Int(bmr(weightKg: profile.weightKg, heightCm: profile.heightCm, age: profile.age, sex: profile.sex))
    }

    /// TDEE for display
    static func tdeeForProfile(_ profile: UserProfile) -> Int {
        let bmrValue = bmr(weightKg: profile.weightKg, heightCm: profile.heightCm, age: profile.age, sex: profile.sex)
        return Int(tdee(bmr: bmrValue, activityLevel: profile.activityLevel))
    }
}
